### ISSUE TITLE
[BUG] Potential Spatial Soft Argmax Bug

### ISSUE THREAD

**Issue by pablovela5620** [kornia/kornia#353]
Hi, I have the following issue. I'm not sure if this is the intended behavior of the `kornia.geometry.spatial_softargmax_2d` but I seem to be having trouble with it. I have the following heatmap. I created this in such a way that the min value is very close to zero and the max value is 1. I made sure to use `spatial_softmax_2d` as stated in the documentation it should be a valid heatmap.
![image](https://user-images.githubusercontent.com/25287427/69586327-dd997c00-0fa7-11ea-9f4e-ef650d7543fe.png)
it has a shape of (1,1,224,224) and when I run spatial_softargmax_2d(heatmap, False) I get a value an output of `tensor([[[46647.1523, 65430.5938]]])` when what I expect to get is something like ~`tensor([[[125., 90.]]])`. Am I missing something?

Thanks for the help!

**Comment by ducha-aiki** [comment]
Hi,

Could you please post a code you used to generate the heatmap, so we can run and check?

**Comment by edgarriba** [comment]
@pablovela5620 check #382. Could you provide a snippet to reproduce this issue so that can be included as test ?

**Comment by pablovela5620** [comment]
Sure, this is what I used to generate the heatmaps from uv coords
![image](https://user-images.githubusercontent.com/25287427/71636346-291cf680-2bf4-11ea-9520-644acad0ad76.png)

```python
def create_single_hm(grid, uv, sigma):
    diff = grid-uv
    norm = torch.norm(diff, dim=3)
    return torch.exp(-(norm)/(sigma**2))
```
```python
def generate_heatmaps_from_uv(uv_coords, hm_size, sigma = torch.tensor(3)):
    '''
    uv_tensor (21x2)
    '''
    heatmaps = []
    mesh_grid = kornia.utils.create_meshgrid(hm_size,hm_size, normalized_coordinates=False)
    for uv in uv_coords:
        heatmaps.append(create_single_hm(mesh_grid, uv, sigma))
    heatmaps = torch.stack(heatmaps).squeeze(1)
    return heatmaps
```

**Comment by borgesa** [comment]
Hi,
Discovered Kornia because of the comment in the DSNTNN repository that I've used in a previous project (thanks, @anibali ). Kornia looks like a very promising repository, well done!

I have a couple comments/questions related to this issue and to the soft-argmax implementation in Kornia. As this became longer than I planned (sorry), I'll split the post by topic.

**1) Multiple function calls w/similar name**
I find it confusing that there are two function calls for "spatial soft-argmax", that do slightly different things:

1.     kornia.geometry.spatial_soft_argmax.spatial_soft_argmax2d
2.     kornia.geometry.dsnt.spatial_softargmax_2d


Currently (1) applies "spatial softmax" to the input before applying (2) to the resulting 'soft_input'.
I would propose to rename one of the functions.

I suppose it depends on how one defines the soft-argmax function:
In [1], soft-argmax is defined explicitly to include the softmax operation (equation 2 in the paper).
This corresponds to implementation (1).

In [2] (the DSNT paper), it is assumed that the input "is a single-channel normalized heatmap" (normalized -> non-negative elements, summing to one). In other words, the input is already assumed to have been normalized.
This corresponds to implementation (2).

**References:**
[1] Human Pose Regression by Combining Indirect Part Detection and Contextual
Information
[2] Numerical Coordinate Regression with Convolutional Neural Networks

**Comment by borgesa** [comment]
**2) Regarding the issue related to usage of 'kornia.geometry.dsnt.spatial_softmax_2d'**
I don't know exactly how @pablovela5620 applies the heatmaps to the 'spatial_softmax_2d'-function, so I might be way off in what follows:

A heatmap generated by the example function 'generate_heatmaps_from_uv' has a single max value of 1. This breaks the assumption that the DSNT input sums to one (assumption made in 'kornia.geometry.spatial_softargmax_2d').

I saw that there is already a function that generates a suitable gaussian:
```
kornia.geometry.dsnt.render_gaussian_2d
```
The function 'kornia.geometry.dsnt.spatial_softargmax_2d' can be tested by applying a gaussian created by that function. Note that 'render_gaussian_2d' seems to cap values to zero at around 1e-15.

To test using gaussians from "generate_heatmaps_from_uv, I believe it would be best to normalize the values s.t. the heatmap values sum to one:
```python
input_ = generate_heatmaps_from_uv(center_point, grid_size, std)
input_normalized = input_ / input_.sum()
output = dsnt.spatial_softargmax_2d(input_normalized)
```

If, on the other hand, softmax is appllied to the heatmap generated by 'generate_heatmaps_from_uv', the output is flattened to such a degree that soft-argmax seemingly yields the center coordinate.
One "dirty fix" in this particular case is:
```python
scaling_factor = 10 # or some other high number
input_ = generate_heatmaps_from_uv(center_point, grid_size, std)
soft_input = dsnt.spatial_softmax_2d(scaling_factor*input_)
output = dsnt.spatial_softargmax_2d(soft_input)
```

**Comment by borgesa** [comment]
**3) Docstring update **
Some documentation related to the dsnt/soft-argmax modules needs to be updated (they still point to 'kornia.contrib', which is a bit confusing).

**4) Implementation plan **
Do you have an estimate of when v0.2.1 will be released/this issue will be closed?

**Comment by anibali** [comment]
What @borgesa said is spot on, and I suspect that they have also found the true cause of the problem @pablovela5620 is having.

The name clash is mainly my fault, since I was the one that brought the important parts of `dsntnn` into Kornia after some initial work had already been done. My thinking at the time was that I wanted to make it easier for people to transition to Kornia from `dsntnn`, but in hindsight that perhaps wasn't the best idea. I wouldn't be against renaming `kornia.geometry.dsnt.spatial_softargmax_2d` to something like `kornia.geometry.dsnt.expected_coordinates_2d` and deprecating the old name, since technically the function calculates the expectation of X, Y coordinate values using the normalised heatmap as a probability mass function.

**Comment by edgarriba** [comment]
@anibali @pablovela5620 was this issue solved ?

**Comment by anibali** [comment]
@edgarriba Yes, I think so. It seems like it was an issue related to confusing similarly named functions with each other. The naming has since been clarified (`spatial_softargmax_2d` -> `spatial_expectation2d`)

**Comment by edgarriba** [comment]
Great, thanks!
